CONTRIBUTING.md

```diff
@@ -38,3 +38,60 @@ out=$(pwd) genericBuild
 ```
 
 Build logs are written to stdout and artifacts written to the current directory.
+
+## Releasing packages to CHaP
+
+When new versions of the packages are released, they should be included in [CHaP](https://github.com/input-output-hk/cardano-haskell-packages).
+See the CHaP README for [instructions](https://github.com/input-output-hk/cardano-haskell-packages#-from-github).
+
+## Updating dependencies
+
+Our Haskell packages come from two package repositories:
+- Hackage
+- [CHaP](https://github.com/input-output-hk/cardano-haskell-packages) (which is essentially another Hackage)
+
+The "index state" of each repository is pinned to a particular time in
+`cabal.project`.  This tells Cabal to treat the repository "as if" it was
+the specified time, ensuring reproducibility.  If you want to use a package
+version from repository X which was added after the pinned index state
+time, you need to bump the index state for X.  This is not a big deal,
+since all it does is change what packages `cabal` considers to be available
+when doing solving, but it will change what package versions cabal picks
+for the plan, and so will likely result in significant recompilation, and
+potentially some breakage.  That typically just means that we need to fix
+the breakage (and add a lower-bound on the problematic package), or add an
+upper-bound on the problematic package.
+
+Note that `cabal` itself keeps track of what index states it knows about, so
+when you bump the pinned index state you may need call `cabal update` in
+order for `cabal` to be happy.
+
+The Nix code which builds our packages also cares about the index state.
+This is represented by inputs managed by `niv`:
```

Comment:
When would I want `niv` to use a different index-state than the one written in my `cabal.project` file?

If _never_, then can we automate this synchronization step somehow?

Reply:
This is very hard. The crux is that we want to accomodate both nix and not-nix users.

Nix users would not need index-state in `cabal.project` at all, because cabal (as run by nix build) does not access hackage at all but only a local snapshot of it that nix has made.

But developers run `cabal` also outside of a nix build, which does require index-state in cabal.project to have the equivalent effect.

I don't know how to improve this situation :disappointed:

Reply:
I think the question is maybe slightly different. The inputs being managed by `niv` do not _determine_ the index-state being used, that is done by `cabal.project` in all cases. What they do is _provide the information that allows nix to handle that index-state_. So essentially the property is "niv input date >= index-state in cabal.project" to make nix happy. We can't easily automate this because both changes need something explicitly written in the repository.

Reply:
Ah, OK, Michael's explanation helps 👍 

New suggestion:

> Our Nix infrastructure ignores `cabal.project`'s index state and instead uses the inputs specified via `niv:
> ...
> When changing the index-state, you should also update these by running:

Reply:
No, what you wrote isn't right either. Did you read my change to the text? I tried to explain it.

Reply:
FTR: It is impossible to forget to update the `hackage.nix` `niv` input; if your `index-state` in `cabal.project` is newer than the `niv` pin, you get an error message like this:
```
error: Unknown index-state 2022-11-01T00:00:00Z, the latest index-state I know about is 2022-08-18T00:00:00Z. You may need to update to a newer hackage.nix.
```

Reply:
Oh right, that's great :smile: